### ISSUE TITLE
Constrained geometry optimization of excited states by casscf with geomeTRIC

### ISSUE THREAD

**Issue by hochej** [pyscf/pyscf#362]
Hi,

I am trying do to a state-averaged CASSCF constrained geometry optimization in the first excited state using the geomeTRIC package.
For this I slightly modified two lines in the <code>geomopt/geometric_solver.py</code> to pass also the geometric constraints to the <code>kernel</code> function (line 167)

before:

    return kernel(method, assert_convergence, include_ghost, callback,
                  maxsteps, **kwargs)[1]
after:

    return kernel(method, assert_convergence, include_ghost, constraints,
                  callback, maxsteps, **kwargs)[1]

and I also added the electronic state as an argument to the kernel function (line 92)

before:

    def kernel(method, assert_convergence=ASSERT_CONV,
               include_ghost=INCLUDE_GHOST, constraints=None, callback=None,
                maxsteps=100, **kwargs):
           
after: 

     def kernel(method, assert_convergence=ASSERT_CONV,
                include_ghost=INCLUDE_GHOST, constraints=None, callback=None,
                maxsteps=100, state=None, **kwargs):

to be able to pass the electronic state as an argument to the <code>as scanner</code> method in line 115.   

  
before:

    elif getattr(method, 'nuc_grad_method', None):
        g_scanner = method.nuc_grad_method().as_scanner()

after:

    elif getattr(method, 'nuc_grad_method', None):
        if state:
            g_scanner = method.nuc_grad_method().as_scanner(state=state)
        else:
            g_scanner = method.nuc_grad_method().as_scanner()
            

After these minor changes it seems to work fine for small test systems like benzene. However when I am trying to optimize my system of interest (58 atoms) it only works for a minimal basis set like STO-3G. Using a "bigger" basis set like 6-31G, 6-31G* or def2-SVP will always result in an error, but strangely only after 3 to 8 optimization cycles. I got this error message:

`
******** <class 'pyscf.grad.casci.CASCI_GradScanner'> for <class 'pyscf.mcscf.casci.CASCI_Scanner'> ********
max_memory 20000 MB (current use 26650 MB)
Traceback (most recent call last):
  File "./cas_scan.py", line 112, in <module>
    mol_eq = optimize(mc, constraints="constraints.txt", maxsteps=maxsteps, state=state_id, **conv_params)
  File "/home/hochej/software/pyscf/pyscf/geomopt/geometric_solver.py", line 171, in optimize
    callback,maxsteps, **kwargs)[1]
  File "/home/hochej/software/pyscf/pyscf/geomopt/geometric_solver.py", line 143, in kernel
    constraints=constraints, **kwargs)
  File "/home/hochej/.local/lib/python2.7/site-packages/geometric/optimize.py", line 1792, in run_optimizer
    progress = Optimize(coords, M, IC, engine, dirname, params, xyzout)
  File "/home/hochej/.local/lib/python2.7/site-packages/geometric/optimize.py", line 1326, in Optimize
    return optimizer.optimizeGeometry()
  File "/home/hochej/.local/lib/python2.7/site-packages/geometric/optimize.py", line 1290, in optimizeGeometry
    self.calcEnergyForce()
  File "/home/hochej/.local/lib/python2.7/site-packages/geometric/optimize.py", line 996, in calcEnergyForce
    spcalc = self.engine.calc(self.X, self.dirname)
  File "/home/hochej/.local/lib/python2.7/site-packages/geometric/engine.py", line 166, in calc
    result = self.calc_new(coords, dirname)
  File "/home/hochej/software/pyscf/pyscf/geomopt/geometric_solver.py", line 83, in calc_new
    energy, gradients = g_scanner(mol)
  File "/home/hochej/software/pyscf/pyscf/grad/casci.py", line 231, in __call__
    de = self.kernel(ci=ci, state=state, **kwargs)
  File "/home/hochej/software/pyscf/pyscf/grad/casci.py", line 287, in kernel
    self.de = kernel(self.base, mo_coeff, ci, atmlst, mf_grad, log)
  File "/home/hochej/software/pyscf/pyscf/grad/casci.py", line 155, in kernel
    de[k,i] += numpy.einsum('ijkl,il,kj', eri1tmp, hf_dm1[p0:p1], zvec_ao[q0:q1])
  File "/home/hochej/software/anaconda2/lib/python2.7/site-packages/numpy/core/einsumfunc.py", line 1346, in einsum
    return c_einsum(*operands, **kwargs)
ValueError: operands could not be broadcast together with remapped shapes [original->remapped]: (9,2,337,337)->(9,2,337,337) (9,337)->(9,newaxis,newaxis,337) (2,337)->(337,2,newaxis)
`

I ran this simulation on a node with 28 physical cores and 192 GB ram. Since I get this error only when using larger basis sets in conjunction with relatively large molecules, I am not sure if this is just a memory issue or if this type of job is not properly implemented in PySCF and it is just lucky that it works in some cases.

My input looks like this:

    import time
    from pyscf import gto, scf, mcscf, grad
    from pyscf import lib
    import qusim as qs
    from pyscf import fci
    from pyscf.geomopt.geometric_solver import optimize

    starttime = time.time()

    ### SETTINGS ######################
    ncpu = 56 
    xyz_name = "geometry.xyz"
    basis_set = "6-31G"
    xyz_unit = "Angstrom"
    charge = 0 
    spin = 0 
    nstates = 2 
    state_id = 1 
    nelec = 4 
    norb = 4 
    verbose = 4 
    maxsteps = 500 
    memory = 20000
    ####################################

    lib.num_threads(ncpu)
    print "Number of processors:", lib.num_threads()

    ### INITIALIZATION ##################
    mol = gto.Mole()
    molecule, charge, spin = qs.read_molecule(xyz_name)
    mol = gto.Mole(atom=molecule, charge=charge, spin=spin).build()
    mol.charge = charge
    mol.spin = spin
    mol.unit = xyz_unit
    mol.verbose = verbose
    mol.basis = basis_set
    mol.max_memory = memory
    mol.build()
    #####################################

    ### RHF REFERENCE CALCULATION #######
    # Initialize the reference HF calculation
    mf = scf.RHF(mol).run()
    #####################################
                                                                                                                     
    ### STATE AVERAGED CASSCF ###########
    mc = mcscf.CASSCF(mf, norb, nelec)
    # All core orbitals are frozen and not optimized
    mc.frozen = mc.ncore
    mc.fcisolver = fci.solver(mol, singlet=True)
    mix_coeff = [1./nstates for _ in range(nstates)]
    mc = mcscf.state_average(mc, mix_coeff)
    mo = mc.mo_coeff
    #####################################

    ### CASCI CALCULATION ################
    mc = mcscf.CASCI(mf, norb, nelec)
    mc.fcisolver = fci.solver(mol, singlet=True)
    mc.fcisolver.nstates = nstates
    # Set optimized orbitals of CASSCF calculations
    emc = mc.casci(mo)[0]
    mc.run()
    ####################################

    ### OPTIMIZATION ################
    conv_params = {
        'convergence_energy': 5e-4,  # Eh
        'convergence_grms': 1e-2,    # Eh/Bohr
        'convergence_gmax': 2.5e-2,  # Eh/Bohr
        'convergence_drms': 1.2e-1,  # Angstrom
        'convergence_dmax': 1.8e-1,  # Angstrom
    }
    mol_eq = optimize(mc, constraints="constraints.txt", maxsteps=maxsteps, state=state_id, **conv_params)
    ####################################
    
    print "Total Walltime: ", time.time()-starttime, " s"

Do you have any ideas about this? 

Thank you very much in advance! 

Joscha Hoche

**Comment by sunqm** [comment]
Does the geometry optimization work if constraints are not specified?

**Comment by sunqm** [comment]
Which pyscf version were you using? state-averaged casscf/casci geometry optimization was recently fixed in dev branch (issue #335, #340).  Does the geometry optimization work without specifying geometry constraints?

**Comment by hochej** [comment]
I am using the current master branch and also tried now the dev branch. But even without specifying geometry constraints (and the state of interest), I get the same error as before.

**Comment by hochej** [comment]
After getting access to a node with 400 GB memory, I repeated the simulations, and after setting `mol.max_memory` to value like 200000, I was able to fix the problem. Please excuse me for opening a new issue due to my own hardware limitations and many thanks for your help.
